init/luks.go

```diff
@@ -299,6 +301,27 @@ func recoverTokenPassword(volumes chan *luks.Volume, d luks.Device, t luks.Token
 	info("password from %s token #%d does not match", t.Type, t.ID)
 }
 
+func recoverKeyfilePassword(volumes chan *luks.Volume, d luks.Device, checkSlots []int, mappingName string, keyfile string) {
+	password, err := ioutil.ReadFile(keyfile)
```

Comment:
does it mean you are trying to read the password from the file packaged into the initramfs? Is it normal practice?

I would always assume a standard place to store the password file is either EFI partition (next to kernel and initramfs) or an external USB flash device.

Reply:
Yes, a fairly normal practice is to store the password within initramfs which is itself stored in /boot, which itself is encrypted and requires grub (or your bootloader of choice) to unlock. Either because root contains /boot, or in many current cases a separate luksv1 partition (because grub and luksv2 don't really play nice still). The reason for this being a normal approach is it is often the "easiest" since init can easily access files in initramfs.

The file could also be stored in any other encrypted location, assuming it has been unlocked already. The primary use case for the keyfile is to prevent needing to type a password multiple times (once for grub, once for luks to unseal) when using full disk encryption. It can be combined with TPM autounsealing though to potentially allow unattended boot of encrypted volumes that contain the additional keys to unseal the remaining volumes too, but I view that as a security risk personally and require a password entry on every boot to unseal the boot partition. In these cases EFI will contain a single efi executable, the signed grub with all its modules preloaded and the bundled config pointing it to whatever volume holds /boot.

Storing it on an external drive is also an option, but that will require implementing the full keyfile spec to allow pointing to devices as well. I can certainly take a look at how complex that process would be to implement. I wouldn't mind storing the key right on /boot and referencing it directly, I was just starting with the lowest hanging fruit first :)

The primary reason for all this is that TPM is still a minefield on a lot of platforms (ARM for instance, VMs for another) so there is a lot of cases where people are just using secureboot combined with luks to harden security in those cases. It just also happens that fast and lightweight principals of booster coincide with those platforms use cases.

This gives me some stuff to look at more and once I get my systems up to where I can test I can toy with a few configurations as well, thanks!

Reply:
Ok fail points.

Though I would love to have a way to keep a keyfile at EFI or and USB drive as well. And we need to think about how to reference different locations. if initramfs-based key referenced by `rd.luks.key=UUID=path` what would be the configuration for EFI based one? for a USB based one?

Reply:
Speaking of TPM key wrapping clevis is a good library to do it. We can use it to wrap the keyfile with TPM or Yubikey.

https://github.com/anatol/clevis.go

Reply:
> if initramfs-based key referenced by `rd.luks.key=UUID=path` what would be the configuration for EFI based one? for a USB based one?

The official keyflie spec is `rd.luks.key=<UNLOCK UUID>=path:UUID=<STORAGE UUID>` where `:UUID=<STORAGE UUID>` is optional. Technically the unlock uuid is also optional, if not specified it is assumed to be the root/old provided rd.luks device specified. The storage uuid is expected to be the uuid of whatever device is holding the keyfile and is assumed you have provided necessary filesystem drivers to open it.